tests/chainer_tests/optimizers_tests/test_optimizers.py

```diff
@@ -45,4 +47,72 @@ def test_hyperparams(self):
             self.assertEqual(self.get_hyperparam(name), new_value)
 
 
+class WeightSaveHook(object):
+    name = 'WeightSaveHook'
+    call_for_each_param = True
+
+    def __init__(self):
+        self.value = None
+
+    def __call__(self, rule, param):
+        p, g = param.data, param.grad
+        if p is None or g is None:
+            return
+        self.value = np.copy(p)
+
+
+class SimpleChain(chainer.Chain):
+
+    def __init__(self):
+        super(SimpleChain, self).__init__()
+        with self.init_scope():
+            self.w = chainer.Parameter(42, (), 'w')
+
+    def __call__(self, x):
+        return (x - self.w) ** 2
+
+
+@testing.parameterize(*testing.product({
+    'impl': [
+        optimizers.AdaDelta,
+        optimizers.AdaGrad,
+        optimizers.Adam,
+        optimizers.MomentumSGD,
+        optimizers.NesterovAG,
+        optimizers.RMSprop,
+        optimizers.RMSpropGraves,
+        optimizers.SGD,
+        optimizers.SMORMS3,
+    ]
+}))
+class TestOptimizerHooks(unittest.TestCase):
+
+    def setUp(self):
+        self.target = SimpleChain()
+
+    def create(self, *args, **kwargs):
+        self.optimizer = self.impl(*args, **kwargs)
+        self.optimizer.setup(self.target)
+
+    def get_hyperparam(self, name):
+        return getattr(self.target.w.update_rule.hyperparam, name)
+
+    def test_hooks(self):
+        w_pre = np.copy(self.target.w.data)
+        h_pre = WeightSaveHook()
+        h_post = WeightSaveHook()
+        self.create()
+        self.optimizer.add_hook(h_pre)
```

Comment:
Current tests do not check the behavior of `timing=pre` case.
I would suggest to put `timing='pre'` explicitly here and check `timing=auto` case as an another test.

Reply:
Ok, will do.